File: dart/test/unicode_rule_label_classifier_test.dart
// FUTURE-PARITY-BACKLOG.10.5.0.2.0 — generated Dart rule-label primitives.

import 'dart:convert';
import 'dart:io';

import 'package:linkedspec_dart/src/parser/unicode_rule_label.dart';
import 'package:test/test.dart';

void main() {
  final contract =
      jsonDecode(
            File(
              '../capability_conformance/unicode_rule_label_contract.json',
            ).readAsStringSync(),
          )
          as Map<String, Object?>;
  final ranges = (contract['xid_continue_ranges']! as List<Object?>)
      .cast<List<Object?>>();
  final positives = (contract['positive_fixtures']! as List<Object?>)
      .cast<Map<String, Object?>>();
  final negatives = (contract['negative_fixtures']! as List<Object?>)
      .cast<Map<String, Object?>>();
  final distinct = (contract['distinct_fixtures']! as List<Object?>)
      .cast<Map<String, Object?>>();

  test('generated metadata and all range boundaries match the contract', () {
    expect(unicodeRuleLabelContractId, contract['contract_id']);
    expect(unicodeRuleLabelVersion, contract['unicode_version']);
    expect(unicodeRuleLabelDataSha256, contract['data_sha256']);
    expect(unicodeRuleLabelRangeCount, ranges.length);

    for (final range in ranges) {
      final start = int.parse(range[0]! as String, radix: 16);
      final end = int.parse(range[1]! as String, radix: 16);
      expect(
        isRuleLabelCodePoint(start),
        isTrue,
        reason: 'range start U+${range[0]}',
      );
      expect(
        isRuleLabelCodePoint(end),
        isTrue,
        reason: 'range end U+${range[1]}',
      );
    }
    expect(isRuleLabelCodePoint(-1), isFalse);
    expect(isRuleLabelCodePoint(0xD800), isFalse);
    expect(isRuleLabelCodePoint(0x110000), isFalse);
  });

  test('complete-label validation matches every neutral fixture', () {
    for (final fixture in positives) {
      final label = fixture['label']! as String;
      expect(isRuleLabel(label), isTrue, reason: fixture['id']! as String);
    }
    for (final fixture in negatives) {
      final label = fixture['label']! as String;
      expect(isRuleLabel(label), isFalse, reason: fixture['id']! as String);
    }
    for (final fixture in distinct) {
      final left = fixture['left']! as String;
      final right = fixture['right']! as String;
      expect(isRuleLabel(left), isTrue);
      expect(isRuleLabel(right), isTrue);
      expect(left, isNot(equals(right)), reason: fixture['id']! as String);
    }
  });

  test('longest-prefix scanning preserves scalar and UTF-16 boundaries', () {
    for (final fixture in positives) {
      final label = fixture['label']! as String;
      final scan = takeRuleLabelPrefix('$label[2]');
      expect(scan, isNotNull, reason: fixture['id']! as String);
      expect(scan!.label, label);
      expect(scan.remainder, '[2]');
    }

    final partial = takeRuleLabelPrefix('Top😀Rule');
    expect(partial!.label, 'Top');
    expect(partial.remainder, '😀Rule');
    expect(takeRuleLabelPrefix('😀Top'), isNull);
    expect(takeRuleLabelPrefix(''), isNull);
  });
}

File: julia/test/unicode_rule_label_classifier_test.jl
# FUTURE-PARITY-BACKLOG.10.6.1.1 — generated Julia rule-label primitives.

const JULIA_UNICODE_RULE_LABEL_CONTRACT = JSON3.read(
    read(
        joinpath(REPO_ROOT, "capability_conformance", "unicode_rule_label_contract.json"),
        String,
    ),
    Dict{String,Any},
)

@testset "generated metadata and all range boundaries match the contract" begin
    contract = JULIA_UNICODE_RULE_LABEL_CONTRACT
    ranges = contract["xid_continue_ranges"]
    @test LinkedSpecJulia.UNICODE_RULE_LABEL_CONTRACT_ID == contract["contract_id"]
    @test LinkedSpecJulia.UNICODE_RULE_LABEL_VERSION == contract["unicode_version"]
    @test LinkedSpecJulia.UNICODE_RULE_LABEL_DATA_SHA256 == contract["data_sha256"]
    @test LinkedSpecJulia.UNICODE_RULE_LABEL_RANGE_COUNT == length(ranges)

    for range in ranges
        start = parse(Int, String(range[1]); base = 16)
        stop = parse(Int, String(range[2]); base = 16)
        @test LinkedSpecJulia.is_rule_label_codepoint(start)
        @test LinkedSpecJulia.is_rule_label_codepoint(stop)
    end
    @test !LinkedSpecJulia.is_rule_label_codepoint(-1)
    @test !LinkedSpecJulia.is_rule_label_codepoint(0xD800)
    @test !LinkedSpecJulia.is_rule_label_codepoint(0x110000)
end

@testset "complete-label validation matches every neutral fixture" begin
    contract = JULIA_UNICODE_RULE_LABEL_CONTRACT
    for fixture in contract["positive_fixtures"]
        @test LinkedSpecJulia.is_rule_label(String(fixture["label"]))
    end
    for fixture in contract["negative_fixtures"]
        @test !LinkedSpecJulia.is_rule_label(String(fixture["label"]))
    end
    for fixture in contract["distinct_fixtures"]
        left = String(fixture["left"])
        right = String(fixture["right"])
        @test LinkedSpecJulia.is_rule_label(left)
        @test LinkedSpecJulia.is_rule_label(right)
        @test left != right
    end
end

@testset "longest-prefix scanning preserves scalar and Julia string boundaries" begin
    for fixture in JULIA_UNICODE_RULE_LABEL_CONTRACT["positive_fixtures"]
        label = String(fixture["label"])
        scan = LinkedSpecJulia.take_rule_label_prefix(string(label, "[2]"))
        @test scan !== nothing
        @test scan.label == label
        @test scan.remainder == "[2]"
    end

    partial = LinkedSpecJulia.take_rule_label_prefix("Top😀Rule")
    @test partial !== nothing
    @test partial.label == "Top"
    @test partial.remainder == "😀Rule"
    @test LinkedSpecJulia.take_rule_label_prefix("😀Top") === nothing
    @test LinkedSpecJulia.take_rule_label_prefix("") === nothing
end
